### ISSUE TITLE
Built in data normalization

### ISSUE THREAD

**Issue by lodenrogue** [BrainJS/brain.js#25]
Any chance we could get built in data normalization?

**Comment by robertleeplummerjr** [comment]
How would you like it normalized?

**Comment by lodenrogue** [comment]
To fit the brain input/output. If my variables are greater than 1 I want it to normalize the data to be within a range of 0 to 1.

**Comment by robertleeplummerjr** [comment]
I've been thinking about this for a bit, currently there is something like this built into brain.js.  Consider the current README.md:

``` js
var net = new brain.NeuralNetwork();

net.train([{input: { r: 0.03, g: 0.7, b: 0.5 }, output: { black: 1 }},
           {input: { r: 0.16, g: 0.09, b: 0.2 }, output: { white: 1 }},
           {input: { r: 0.5, g: 0.5, b: 1.0 }, output: { white: 1 }}]);

var output = net.run({ r: 1, g: 0.4, b: 0 });  // { white: 0.99, black: 0.002 }
```

Here `r`, `g`, and `b` are normalized to the layers associated with them.  This may not be exactly what you want, especially if you have thousands or more types.  Another way you could do it, say if you had an object with the properties that you'd like to normalize:

``` js
var keys = mySuperObject.keys();
var normalizedValues = keys.map(function(key, i) {
  return i / keys.length;
});
```

**Comment by lodenrogue** [comment]
Lets say I have 10 mapped keys. If the value of one of those keys is 30 and I'm dividing that by 10 then the returned value is 3. That's not between 0 and 1. 

A way I've found is iterating through your csv and finding the highest value. Then going back through the list again and dividing all values by this max value. This becomes slow when you have many values. 

Maybe we can use some variation of the sigmoid function and tailor it to our needs?

**Comment by robertleeplummerjr** [comment]
There is probably a better way, but try something like this:

``` js
var values = [];
for (var i = 0, max = 10000; i < max; i++) {
  values.push(i);
}

var normalizedInputValues = [];
values.forEach(function(value, index) {
  normalizedInputValues.push(index / values.length);
});

for (i = 0, max = normalizedInputValues.length; i < max; i++) {
  if (normalizedInputValues[i] > 1) throw new Error('This should never happen, but there is a value above 1 here at index of ' + i);
}

//have fun
```

**Comment by robertleeplummerjr** [comment]
In the above mentioned:
```js
var net = new brain.NeuralNetwork();

net.train([{input: { r: 0.03, g: 0.7, b: 0.5 }, output: { black: 1 }},
           {input: { r: 0.16, g: 0.09, b: 0.2 }, output: { white: 1 }},
           {input: { r: 0.5, g: 0.5, b: 1.0 }, output: { white: 1 }}]);

var output = net.run({ r: 1, g: 0.4, b: 0 });  // { white: 0.99, black: 0.002 }
```

The data _is sort of_ normalized, but by layer, would this work for you?

```js
var net = new brain.NeuralNetwork();

net.train([
  {input: { one: 0.03, two: 0.7, ... thirty: 0.5 }, output: { isValid: 1 }},
  {input: { one: 1, two: 0.02, ... thirty: 0.33 }, output: { isValid: 1 }},
  {input: { one: 0.3, two: 0.2, ... thirty: 0.2 }, output: { isValid: 0 }}
]);
var output = net.run({ one: 0.03, two: 0.8, ... thirty: 0.6 });  // { isValid: .89 }
```

Does this help?

**Comment by robertleeplummerjr** [comment]
The other option at this point could be to use a recurrent neural net, such as an LSTM, but this is more for data translation and less about math.

**Comment by lodenrogue** [comment]
I've found another solution. Thank you for your attention.

**Comment by robertleeplummerjr** [comment]
Care to share?  We always want to improve for usefulness.

**Comment by lodenrogue** [comment]
I went with a different library: scikit-learn

**Comment by robertleeplummerjr** [comment]
ty for the pointer!
